pkg/scheduler/scheduler.go

```diff
@@ -299,15 +300,19 @@ func (sched *Scheduler) Run(ctx context.Context) {
 
 // recordSchedulingFailure records an event for the pod that indicates the
 // pod has failed to schedule. Also, update the pod condition and nominated node name if set.
-func (sched *Scheduler) recordSchedulingFailure(fwk framework.Framework, podInfo *framework.QueuedPodInfo, err error, reason string, nominatedNode string) {
+func (sched *Scheduler) recordSchedulingFailure(fwk framework.Framework, podInfo *framework.QueuedPodInfo, err error, reason string, nominatedNode *string) {
 	sched.Error(podInfo, err)
 
 	// Update the scheduling queue with the nominated pod information. Without
 	// this, there would be a race condition between the next scheduling cycle
 	// and the time the scheduler receives a Pod Update for the nominated pod.
 	// Here we check for nil only for tests.
 	if sched.SchedulingQueue != nil {
-		sched.SchedulingQueue.AddNominatedPod(podInfo.PodInfo, nominatedNode)
+		var nnn string
```

Comment:
It looks like `nil` and empty string have the same result here. Why use a pointer?

Reply:
Ah, it's more about `updatePod` below.

However, it looks like `AddNominatedPod` looks at the existing `nominatedNode` in the status when you pass empty string. This means that the nominated node name is not actually cleared from the cache.

Reply:
> Ah, it's more about updatePod below.

Yes.

> However, it looks like AddNominatedPod looks at the existing nominatedNode in the status when you pass empty string.

Not quite. The entry in internal nominator implementation will be cleared first in all cases:

https://github.com/kubernetes/kubernetes/blob/4cc1abf2474efee4e358626856b5fc8a5675baed/pkg/scheduler/internal/queue/scheduling_queue.go#L834-L845

Reply:
I think the problem is this line:

https://github.com/kubernetes/kubernetes/blob/4cc1abf2474efee4e358626856b5fc8a5675baed/pkg/scheduler/internal/queue/scheduling_queue.go#L841